Translated xml configuration to java configuration

muff1nman opened the pull request:
Signed-off-by: Andrew DeMaria <lostonamountain@gmail.com>

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/service/metadata/DefaultMetaDataParser.java:
```diff
@@ -32,6 +33,7 @@
  * @author Sindre Mehus
  */
 @Service
+@Order(1000)
```
Can you please explain (maybe in the commit message) why this one is 1000, and other is 100 and an other other is 0 ?

muff1nman replied:
Just to give a bit of space between to allow for extensibility.

muff1nman replied:
Basically we can add more additional MetaDataParsers and be able to order them between (or before or after) any of the existing ones.

randomnicode replied:
Maybe 0,100,200 makes more sense? Why 1000? Do we expect more parsers between FFMPEG and Default?

Also for additional consideration:
- If additional parsers are added, there are code changes anyway, and the authors can reorder again as they like. People will need to reorder some stuff with more and more parsers anyway. Consider:
  - Author 1 needs to add a Parser1 after FFMPEG and so he naturally uses 101.
  - Author 2 comes along and wants to add the next parser after FFMPEG and before Parser1, he'll _have_ to change order of either FFMPEG or Parser1, Either way it's a code change in another Parser.

Keeping the order linear makes it apparent what is going on.

muff1nman replied:
Sure I'll change it to be linear.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/Application.java:
```diff
@@ -36,10 +35,17 @@
+        "org.airsonic.player.security",
+        "org.airsonic.player.service",
+        "org.airsonic.player.spring",
+        "org.airsonic.player.validator"})
```
Why not replace this with @SpringBootApplication?

It will autoconfigure the ComponentScans. Are we trying to protect some packages from scanning?

muff1nman replied:
I'm not necessarily opposed to `@SpringBootApplication` but for now this keeps the functionality as close as to what it was before. Changing to `@SpringBootApplication` is out of scope for this PR.

randomnicode replied:
SpringBootApplication is already annotated, like a few lines above this comment, I'm not seeing how it is out-of-scope? It's part of the class declaration.

Removing this will also reduce your PR LoC footprint. The only reason you'd have this here would be if you were trying to protect some packages from scanning, which I don't think we have any of.

muff1nman replied:
A separate PR should be considered for component scanning the whole package tree.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/LoggingExceptionResolver.java:
```diff
@@ -4,12 +4,14 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+@Component
```
I would maybe give this an Order ?
(even though it implements Ordered)

muff1nman replied:
Also out of scope. This PR is soley about removing the XML files and keeping the beans as close as possible to how they were defined before.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @Bean
+    public ResourceBundleMessageSource messageSource() {
```
This may be autoinstantiated by putting in a property in the application.properties:
```
spring.messages.basename=org.airsonic.player.i18n.ResourceBundle
```
Do not need a bean here.

muff1nman replied:
Let's change this in a separate PR.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @Bean
+    public CustomThemeSource themeSource(SettingsService settingsService) {
```
Why not autoinstantiate this bean by annotating @Component to the class definition of the bean itself?
Autowire the settingsService and the basename to the provided value.

muff1nman replied:
We can go back and clean it up in a separate PR.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/Application.java:
```diff
@@ -36,10 +35,17 @@
-        "classpath:/applicationContext-cache.xml",
-        "classpath:/applicationContext-sonos.xml",
-        "classpath:/servlet.xml"})
+@EnableWebMvc
```
This annotation should likely be on the ServletConfiguration file

muff1nman replied:
Agreed.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+        handlerMapping.setAlwaysUseFullPath(true);
+        Properties properties = new Properties();
+        properties.put("/podcast/**", podcastController);
+        handlerMapping.setMappings(properties);
```
set a mapping handling order here please. Otherwise you're leaving it to chance and the ResourceHandler might usurp the request with any changes or upgrades.
```
// Default is intmax/lowestprecedence, so need to set a higher priority than
// ResourceHttpRequestHandler (which is intmax-1). Otherwise, that will intercept every request before it gets here
  handlerMapping.setOrder(Integer.MAX_VALUE - 2);
```

Also consider renaming it to podcastUrlMapping, given that is literally the only thing it maps (everything maps to jsps or resources)?

muff1nman replied:
Changed the name. I'm leaving the order unspecified like it was in the previous xml config. A separate PR can change it.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.Properties;
+
+@Configuration
+public class ServletConfiguration {
```
this should probably implement WebMvcConfigurer and be annotated by  @EnableWebMvc

muff1nman replied:
No overriden methods so no need for now.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @Bean
+    public ThemeResolver themeResolver(SecurityService securityService, SettingsService settingsService) {
```
this should be autoinstantiated by adding a @Component  annotation directly on the class and autowiring the dependencies. The  bean instantiation here is unnecessary.

muff1nman replied:
We can go through and do some cleanups like that later. For now just wanted to keep the same logical grouping.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+        return resourceMessageSource;
+    }
+
+    @Bean
```
Affix the correct name of the bean by the powers that be, so if someone dares changes it upstream, we won't have a care in the world: ```UiApplicationContextUtils.THEME_SOURCE_BEAN_NAME```

randomnicode replied:
Or (just remove it)

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+        resolver.setPrefix("/WEB-INF/jsp/");
+        resolver.setSuffix(".jsp");
+        return resolver;
+    }
```
I don't see a resource mapping anywhere.
Any idea where the static resources are getting served from?

Just to make sure it's explicitly defined, do you want to maybe add a resource handler to the register for the static content? You'll need to implement the WebMvcConfigurer class (or its equivalent in Spring Boot 1.5)
```
public void addResourceHandlers(final ResourceHandlerRegistry registry) {
        registry.addResourceHandler("/**").addResourceLocations("/");
}
```

Somewhere in the future I'd like to localize and consolidate all our resources in a subfolder "resources" instead of polluting the root so we can lock down these routes.

muff1nman replied:
Lets address this in a separate PR. Resource handling could use a freshening.

randomnicode commented on airsonic-main/src/test/java/org/airsonic/player/TestCaseUtils.java:
```diff
@@ -90,17 +90,6 @@ public static Integer recordsInTable(String tableName, DaoHelper daoHelper) {
   }
 
 
-  public static ApplicationContext loadSpringApplicationContext(String baseResources) {
```
👍

randomnicode commented on airsonic-main/src/test/java/org/airsonic/player/TestCaseUtils.java:
```diff
@@ -90,17 +90,6 @@ public static Integer recordsInTable(String tableName, DaoHelper daoHelper) {
-    };
-    return new ClassPathXmlApplicationContext(configLocations);
-  }
-
```
👍

randomnicode commented on airsonic-main/src/test/java/org/airsonic/player/service/MediaScannerServiceTestCase.java:
```diff
@@ -161,12 +153,15 @@ public void testScanLibrary() {
         File artistDir = temporaryFolder.newFolder(directoryName);
         File musicFile = artistDir.toPath().resolve(fileName).toFile();
-        IOUtils.copy(resource.getInputStream(), new FileOutputStream(musicFile));
+        IOUtils.copy(resource, new FileOutputStream(musicFile));
```
Maybe just use Google's Resources class, much shorter, and won't need that assertion:
```
Resources.copy(Resources.getResource("MEDIAS/piano.mp3"), new FileOutputStream(musicFile));
```

muff1nman replied:
Separate PR.

randomnicode commented on airsonic-main/src/test/java/org/airsonic/player/service/metadata/MetaDataFactoryTestCase.java:
```diff
@@ -0,0 +1,86 @@
+                TestCaseUtils.cleanAirsonicHomeForTest();
+                File dbDirectory = new File(homeParent, "/db");
+                FileUtils.forceMkdir(dbDirectory);
+                org.airsonic.player.util.FileUtils.copyResourcesRecursively(getClass().getResource("/db/pre-liquibase/db"), new File(homeParent));
```
I think this creates a minor bug.

The default DB URL pointer will be to: `/airsonic_test_temp_folder/db/airsonic`
But your DB is located at: `/airsonic_test_temp_folder/db/libresonic` per the files you copied
Either set a global db url param with a different jdbc url, or rename the files you're copying

muff1nman replied:
Ahh good catch. I copied the classrule from the wrong class.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/DatabaseConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+        dataSource.setUrl(SettingsService.getDefaultJDBCUrl());
+        dataSource.setUsername("sa");
+        dataSource.setPassword("");
+        return dataSource;
```
Can use DataBuilder here also

randomnicode commented on airsonic-main/src/test/resources/applicationContext-testdb.xml:
```diff
@@ -1,15 +0,0 @@
-        <property name="username" value="sa" />
-        <property name="password" value="" />
-    </bean>
-</beans>
```
Why not delete the mockSonos xml too?

muff1nman replied:
good call.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/SonosConfiguration.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Bean
+    public SonosHelper sonosHelper() {
+        return new SonosHelper();
```
This is likely a bug. You've instantiated and returned a blank SonosHelper here with no dependencies wired inside.

You actually don't need to, SonosHelper is already annotated with a component (@Service) so it will autoinstantiate, therefore this bean is not needed

muff1nman replied:
Done.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/SonosConfiguration.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Bean
+    public SonosFaultInterceptor sonosFaultInterceptor() {
+        return new SonosFaultInterceptor();
```
These are ideal targets for autoinstantiation. Singleton? Contextual? No dependencies? Easy win.

Annotate the class and remove the bean from here.

muff1nman replied:
Done.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/spring/DatabaseConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Bean
+    public String userTableQuote(@Value("${DatabaseUsertableQuote:}") String value) {
```
This is a silly bean. I would consider removing it. It's literally a string taken from the environment.

In the bean below encode the `@Value` for `userTableQuote` directly simlar to its previous argument (additionally I think you may also need to annotate the value in UserDao).

muff1nman replied:
It reduces duplication of the @value annotation and has a purpose.

fxthomas commented:
I'm currently getting some warnings when starting this (from e775405aa3dbc69eb33b3a8323fdc1f532aadd1e):

```
$ java -Dairsonic.home=.run -Dserver.port=4040 -jar airsonic-main/target/airsonic.war
...snip...
2019-11-04 21:43:52.357  INFO --- org.airsonic.player.Application          : Started Application in 14.261 seconds (JVM running for 14.828)
2019-11-04 21:44:08.214  INFO --- o.a.p.f.BootstrapVerificationFilter      : Servlet container: Apache Tomcat/8.5.43
2019-11-04 21:44:08.448  WARN --- o.s.web.servlet.PageNotFound             : No mapping found for HTTP request with URI [/style/default.css] in DispatcherServlet with name 'dispatcherServlet'
2019-11-04 21:44:08.448  WARN --- o.s.web.servlet.PageNotFound             : No mapping found for HTTP request with URI [/script/mousetrap-1.6.0.js] in DispatcherServlet with name 'dispatcherServlet'
2019-11-04 21:44:08.455  WARN --- o.s.web.servlet.PageNotFound             : No mapping found for HTTP request with URI [/script/keyboard_shortcuts.js] in DispatcherServlet with name 'dispatcherServlet'
2019-11-04 21:44:08.456  WARN --- o.s.web.servlet.PageNotFound             : No mapping found for HTTP request with URI [/script/mediaelement/mediaelementplayer.min.css] in DispatcherServlet with name 'dispatcherServlet'
2019-11-04 21:44:08.499  WARN --- o.s.web.servlet.PageNotFound             : No mapping found for HTTP request with URI [/icons/default_light/logo.png] in DispatcherServlet with name 'dispatcherServlet'
2019-11-04 21:44:08.528  WARN --- o.s.web.servlet.PageNotFound             : No mapping found for HTTP request with URI [/icons/apple-touch-icon.png] in DispatcherServlet with name 'dispatcherServlet'
2019-11-04 21:44:08.529  WARN --- o.s.web.servlet.PageNotFound             : No mapping found for HTTP request with URI [/icons/favicon-32x32.png] in DispatcherServlet with name 'dispatcherServlet'
```

Related, but the login page has no style:

![Screenshot from 2019-11-04 21-45-21](https://user-images.githubusercontent.com/613594/68156602-752d2100-ff4c-11e9-87e3-2c8d6c9c4d35.png)

...and I'm also getting an error just like #1364 when trying to access Airsonic via http://localhost:4040.

randomnicode commented:
Yeah those Dispatcher Servlet errors are two of the comments I left on the PR that were marked resolved or out of scope.

> I don't see a resource mapping anywhere.
Any idea where the static resources are getting served from?

And

> this should probably implement WebMvcConfigurer and be annotated by @EnableWebMvc

We should close this PR. These issues were already addressed in #1364 . It seems just duplicated work and the more errors that are fixed here the more it'll resemble #1364

muff1nman commented:
@fxthomas fixed. I still prefer this changeset over #1364 as there is some unneeded and/or unrelated changes in #1364 that need to be separated. Also it would need to be rebased.

fxthomas commented:
Well, okay then. In any case this one appears to work correctly at first glance too!

randomnicode commented:
> @fxthomas fixed. I still prefer this changeset over #1364 as there is some unneeded and/or unrelated changes in #1364 that need to be separated. Also it would need to be rebased.

Look friend, I don't think you understood what I was saying, or maybe missed the point. This isn't a question of 1 PR needing to be rebased, as `it would need to be rebased` implies. There is an _entire stack_ of PRs based on your **old** commit. _Every_ one of them will need to be rebased on top of each other. _You_ are unnecessarily creating that work. For no good reason that I can surmise. I'm not going to rebase my 6 PRs because you decided to create additional work for everyone at the end for some strange insistence to merge this PR first.

It's probably easier to retrieve your old commit and merge master into it (plus additional commits for whatever changes you want to make) instead of rebasing it so those  stacked PRs _don't_ have to be rebased.  That way you can get to merge your PR for whatever reason first, and at least this way it'll only affect only 1 PR instead of 6, which I might add is being updated _the last_ (and thus should bear the onus of minimizing the heavy churn it's trying to cause).

Additionally, why is it that the PR that's being updated the _last_ should be merged _first_? Why even update this PR when you've clearly seen it's been taken into another PR? Why not just comment on the other PR? Or merge it and submit patches against the new master? The only reason the other PR was even created was you weren't updating this PR and I couldn't really wait for you while other work was paused, especially since my bigger PR was told it needed to be broken up and based on this.

So, in short, I still don't understand the need to merge this first. Let me reiterate: merging this first is useless. What are you going to do with merging this PR? Ship it as a release? What does a smaller diffset of changes (barely) matter? Everything just sits on master anyway. There are no dev builds. So you just accumulate more commits on master, whether it happens in one PR or a series of PRs. It all goes out in one release.

I really don't get what a smaller diffset (barely, this PR is 33 files now btw) accomplishes when reviews have already been done, builds are clearly passing, and a test run seems to work fine.

Why merge an intermediate PR that you updated at the last second and want to merge first that will cause more work for everyone?

fxthomas commented:
I'm pretty much neutral on the differences between the two. I think both PRs are close to be merged, you two need to agree on what we keep.

Effective diff between A (master+merge #1285) and B (master+merge #1285 then #1364):

```diff
diff --git a/airsonic-main/src/main/java/org/airsonic/player/dao/UserDao.java b/airsonic-main/src/main/java/org/airsonic/player/dao/UserDao.java
index db02b3d0..74f495e1 100644
--- a/airsonic-main/src/main/java/org/airsonic/player/dao/UserDao.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/dao/UserDao.java
@@ -24,6 +24,7 @@ import org.airsonic.player.util.StringUtil;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.jdbc.core.RowMapper;
 import org.springframework.stereotype.Repository;
 import org.springframework.transaction.annotation.Transactional;
@@ -71,7 +72,7 @@ public class UserDao extends AbstractDao {
     private final String userTableQuote;
 
     @Autowired
-    public UserDao(String userTableQuote) {
+    public UserDao(@Value("${DatabaseUsertableQuote:}") String userTableQuote) {
         this.userTableQuote = userTableQuote;
     }
 
diff --git a/airsonic-main/src/main/java/org/airsonic/player/spring/DatabaseConfiguration.java b/airsonic-main/src/main/java/org/airsonic/player/spring/DatabaseConfiguration.java
index 6aa1d455..33091191 100644
--- a/airsonic-main/src/main/java/org/airsonic/player/spring/DatabaseConfiguration.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/spring/DatabaseConfiguration.java
@@ -80,15 +80,11 @@ public class DatabaseConfiguration {
         return new File(SettingsService.getAirsonicHome(), "rollback.sql");
     }
 
-    @Bean
-    public String userTableQuote(@Value("${DatabaseUsertableQuote:}") String value) {
-        return value;
-    }
-
     @Bean
     public SpringLiquibase liquibase(DataSource dataSource,
                                      @Value("${DatabaseMysqlMaxlength:512}")
                                      String mysqlVarcharLimit,
+                                     @Value("${DatabaseUsertableQuote:}")
                                      String userTableQuote) {
         SpringLiquibase springLiquibase = new SpringLiquibase();
         springLiquibase.setDataSource(dataSource);
diff --git a/airsonic-main/src/main/java/org/airsonic/player/spring/LoggingExceptionResolver.java b/airsonic-main/src/main/java/org/airsonic/player/spring/LoggingExceptionResolver.java
index fe411eb5..2295135b 100644
--- a/airsonic-main/src/main/java/org/airsonic/player/spring/LoggingExceptionResolver.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/spring/LoggingExceptionResolver.java
@@ -4,6 +4,7 @@ import org.airsonic.player.util.Util;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.core.Ordered;
+import org.springframework.core.annotation.Order;
 import org.springframework.stereotype.Component;
 import org.springframework.web.servlet.HandlerExceptionResolver;
 import org.springframework.web.servlet.ModelAndView;
@@ -12,6 +13,7 @@ import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 @Component
+@Order(Ordered.HIGHEST_PRECEDENCE)
 public class LoggingExceptionResolver implements HandlerExceptionResolver, Ordered {
 
     private static final Logger LOG = LoggerFactory.getLogger(LoggingExceptionResolver.class);
diff --git a/airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java b/airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java
index 6259a5b0..ee6bfb5a 100644
--- a/airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/spring/ServletConfiguration.java
@@ -4,6 +4,7 @@ import org.airsonic.player.controller.PodcastController;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.web.servlet.ViewResolver;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter;
 import org.springframework.web.servlet.handler.SimpleUrlHandlerMapping;
 import org.springframework.web.servlet.view.InternalResourceViewResolver;
 import org.springframework.web.servlet.view.JstlView;
@@ -11,18 +12,7 @@ import org.springframework.web.servlet.view.JstlView;
 import java.util.Properties;
 
 @Configuration
-public class ServletConfiguration {
-
-    @Bean
-    public SimpleUrlHandlerMapping podcastUrlMapping(PodcastController podcastController) {
-        SimpleUrlHandlerMapping handlerMapping = new SimpleUrlHandlerMapping();
-        handlerMapping.setAlwaysUseFullPath(true);
-        Properties properties = new Properties();
-        properties.put("/podcast/**", podcastController);
-        handlerMapping.setMappings(properties);
-        return handlerMapping;
-    }
-
+public class ServletConfiguration extends WebMvcConfigurerAdapter {
     @Bean
     public ViewResolver viewResolver() {
         InternalResourceViewResolver resolver = new InternalResourceViewResolver();
@@ -31,4 +21,22 @@ public class ServletConfiguration {
         resolver.setSuffix(".jsp");
         return resolver;
     }
+
+    @Bean
+    public SimpleUrlHandlerMapping urlMapping(PodcastController podcastController) {
+        SimpleUrlHandlerMapping mapping = new SimpleUrlHandlerMapping();
+
+        // Default is intmax, so need to set a higher priority than
+        // ResourceHttpRequestHandler/ResourceHandlerRegistry (which is intmax-1). Otherwise, that will
+        // intercept every request before it gets here
+        mapping.setOrder(Integer.MAX_VALUE - 2);
+
+        mapping.setAlwaysUseFullPath(true);
+
+        Properties properties = new Properties();
+        properties.put("/podcast/**", podcastController);
+        mapping.setMappings(properties);
+
+        return mapping;
+    }
 }
diff --git a/airsonic-main/src/test/java/org/airsonic/player/api/jukebox/AbstractAirsonicRestApiJukeboxIntTest.java b/airsonic-main/src/test/java/org/airsonic/player/api/jukebox/AbstractAirsonicRestApiJukeboxIntTest.java
index 92180703..55022aa3 100644
--- a/airsonic-main/src/test/java/org/airsonic/player/api/jukebox/AbstractAirsonicRestApiJukeboxIntTest.java
+++ b/airsonic-main/src/test/java/org/airsonic/player/api/jukebox/AbstractAirsonicRestApiJukeboxIntTest.java
@@ -1,6 +1,7 @@
 package org.airsonic.player.api.jukebox;
 
 import org.airsonic.player.TestCaseUtils;
+import org.airsonic.player.controller.SubsonicRESTController;
 import org.airsonic.player.dao.*;
 import org.airsonic.player.domain.*;
 import org.airsonic.player.service.MediaScannerService;
@@ -176,10 +177,7 @@ public abstract class AbstractAirsonicRestApiJukeboxIntTest {
             jsonPath("$.subsonic-response.jukeboxPlaylist.entry[0].suffix").value(mediaFile.getFormat()).match(result);
             jsonPath("$.subsonic-response.jukeboxPlaylist.entry[0].duration").value(mediaFile.getDurationSeconds()).match(result);
             jsonPath("$.subsonic-response.jukeboxPlaylist.entry[0].bitRate").value(mediaFile.getBitRate()).match(result);
-
-            // The path is absolute, we remove the folder source with replace the folder with nothing
-            // and the "/" they stay with substring(1), must be a method in mediaFile for method to obtain the relative path is same every where.
-            jsonPath("$.subsonic-response.jukeboxPlaylist.entry[0].path").value(mediaFile.getPath().replace(mediaFile.getFolder(), "").substring(1)).match(result);
+            jsonPath("$.subsonic-response.jukeboxPlaylist.entry[0].path").value(SubsonicRESTController.getRelativePath(mediaFile, settingsService)).match(result);
             jsonPath("$.subsonic-response.jukeboxPlaylist.entry[0].isVideo").value(mediaFile.isVideo()).match(result);
             jsonPath("$.subsonic-response.jukeboxPlaylist.entry[0].playCount").isNumber().match(result);
             jsonPath("$.subsonic-response.jukeboxPlaylist.entry[0].created").value(convertDateToString(mediaFile.getCreated())).match(result);
diff --git a/airsonic-main/src/test/java/org/airsonic/player/service/MediaScannerServiceTestCase.java b/airsonic-main/src/test/java/org/airsonic/player/service/MediaScannerServiceTestCase.java
index 55f8c076..22dbb872 100644
--- a/airsonic-main/src/test/java/org/airsonic/player/service/MediaScannerServiceTestCase.java
+++ b/airsonic-main/src/test/java/org/airsonic/player/service/MediaScannerServiceTestCase.java
@@ -3,6 +3,7 @@ package org.airsonic.player.service;
 import com.codahale.metrics.ConsoleReporter;
 import com.codahale.metrics.MetricRegistry;
 import com.codahale.metrics.Timer;
+import com.google.common.io.Resources;
 import org.airsonic.player.TestCaseUtils;
 import org.airsonic.player.dao.*;
 import org.airsonic.player.domain.Album;
@@ -11,7 +12,6 @@ import org.airsonic.player.domain.MediaFile;
 import org.airsonic.player.domain.MusicFolder;
 import org.airsonic.player.util.HomeRule;
 import org.airsonic.player.util.MusicFolderTestData;
-import org.apache.commons.io.IOUtils;
 import org.junit.Assert;
 import org.junit.ClassRule;
 import org.junit.Rule;
@@ -153,15 +153,11 @@ public class MediaScannerServiceTestCase {
 
     @Test
     public void testSpecialCharactersInFilename() throws Exception {
-        InputStream resource = MediaScannerServiceTestCase.class
-                .getClassLoader()
-                .getResourceAsStream("MEDIAS/piano.mp3");
-        assert resource != null;
         String directoryName = "Muff1nman\u2019s \uFF0FMusic";
         String fileName = "Muff1nman\u2019s\uFF0FPiano.mp3";
         File artistDir = temporaryFolder.newFolder(directoryName);
         File musicFile = artistDir.toPath().resolve(fileName).toFile();
-        IOUtils.copy(resource, new FileOutputStream(musicFile));
+        Resources.copy(Resources.getResource("MEDIAS/piano.mp3"), new FileOutputStream(musicFile));
 
         MusicFolder musicFolder = new MusicFolder(1, temporaryFolder.getRoot(), "Music", true, new Date());
         musicFolderDao.createMusicFolder(musicFolder);
diff --git a/airsonic-main/src/test/java/org/airsonic/player/service/metadata/MetaDataFactoryTestCase.java b/airsonic-main/src/test/java/org/airsonic/player/service/metadata/MetaDataFactoryTestCase.java
index 4e5c25b3..53794a24 100644
--- a/airsonic-main/src/test/java/org/airsonic/player/service/metadata/MetaDataFactoryTestCase.java
+++ b/airsonic-main/src/test/java/org/airsonic/player/service/metadata/MetaDataFactoryTestCase.java
@@ -1,7 +1,9 @@
 package org.airsonic.player.service.metadata;
 
+import org.airsonic.player.TestCaseUtils;
 import org.airsonic.player.service.SettingsService;
 import org.airsonic.player.util.HomeRule;
+import org.apache.commons.io.FileUtils;
 import org.junit.*;
 import org.junit.rules.TemporaryFolder;
 import org.junit.runner.Description;
@@ -24,12 +26,23 @@ public class MetaDataFactoryTestCase {
 
     @ClassRule
     public static final SpringClassRule classRule = new SpringClassRule() {
-        HomeRule homeRule = new HomeRule();
+        HomeRule airsonicRule = new HomeRule() {
+            @Override
+            protected void before() throws Throwable {
+                super.before();
+                String homeParent = TestCaseUtils.airsonicHomePathForTest();
+                System.setProperty("airsonic.home", TestCaseUtils.airsonicHomePathForTest());
+                TestCaseUtils.cleanAirsonicHomeForTest();
+                File dbDirectory = new File(homeParent, "/db");
+                FileUtils.forceMkdir(dbDirectory);
+                org.airsonic.player.util.FileUtils.copyResourcesRecursively(getClass().getResource("/db/pre-liquibase/db"), new File(homeParent));
+            }
+        };
 
         @Override
         public Statement apply(Statement base, Description description) {
             Statement spring = super.apply(base, description);
-            return homeRule.apply(spring, description);
+            return airsonicRule.apply(spring, description);
         }
     };
```

muff1nman commented:
> Look friend, I don't think you understood what I was saying, or maybe missed the point. This isn't a question of 1 PR needing to be rebased, as it would need to be rebased implies. There is an entire stack of PRs based on your old commit. Every one of them will need to be rebased on top of each other. You are unnecessarily creating that work. For no good reason that I can surmise. I'm not going to rebase my 6 PRs because you decided to create additional work for everyone at the end for some strange insistence to merge this PR first.

I understand the implications. The work is the sunk cost here. Just because you stacked those PRs on top of one another will not bias on if/how/when the base PR(s) get merged. I'm not trying to be a jerk, but rather this is a result of this community being very iterative on doing work. Whether that's good/bad/not inline with other FOSS communities can be debated separately as its irrelevant. 
 
> It's probably easier to retrieve your old commit and merge master into it (plus additional commits for whatever changes you want to make) instead of rebasing it so those stacked PRs don't have to be rebased. That way you can get to merge your PR for whatever reason first, and at least this way it'll only affect only 1 PR instead of 6, which I might add is being updated the last (and thus should bear the onus of minimizing the heavy churn it's trying to cause).

You're assuming here that your PRs are going to get merged at all. In addition if your PRs are nicely contained it should be straightforward to get them updated with either a rebase or merge. 

> Additionally, why is it that the PR that's being updated the last should be merged first? Why even update this PR when you've clearly seen it's been taken into another PR? Why not just comment on the other PR? Or merge it and submit patches against the new master? The only reason the other PR was even created was you weren't updating this PR and I couldn't really wait for you while other work was paused, especially since my bigger PR was told it needed to be broken up and based on this.

I've updated this PR because your PR contains changes that I do not want to merge (either in the same PR, or even at all). In addition you've given me the impression you're not going to squash your changes. Am I being a meanie for making people break up their PRs into bite size chunks? Probably. However it is done for some very good reasons. Mainly 1. Allows easily reverting. 2. Makes for readable git history. 3. Makes bisecting issues much easier. 

> So, in short, I still don't understand the need to merge this first. Let me reiterate: merging this first is useless. What are you going to do with merging this PR? Ship it as a release? What does a smaller diffset of changes matter? Everything just sits on master anyway. There are no dev builds. So you just accumulate more commits on master, whether it happens in one PR or a series of PRs. It all goes out in one release.

We are going to work incrementally. Maybe this is all we are going to work on for spring boot right now. That sounds crazy but you have to realize there is no product manager, no full-time dev team committed to seeing a large change through end to end. Maybe you or I just vanish tomorrow and to never finish and never help the end users with bugs that surface as a result of a half-done large spring boot migration.  Therefore we work incrementally and cautiously. 

> a test run seems to work fine.

Confirms what I just said above. We need to build confidence through smaller changes.

> cause more work for everyone?

You mean cause more work for you? PRs that are not clean, self contained and squashed cause more work for everyone else. 

So look here. At the end of the day I like your intentions. You want to help update spring, want to help with other bits of the application that are in dire need of attention. Great stuff. However, you have to play by the rules here. What rules?! Where are they written down?! you might say. And this is the beauty of this open source community. These rules change over time and exist between the current members of the team. When Sindre was in charge - there were a completely separate set of rules compared to when Eugene was in change, compared to when it was just biconou, jooola and I working on it or compared to now when we arguably have the most amount of concurrent contributors. To keep updated rules documented takes time and effort that, until recently, none of us had.

So what do we do without a pair of rules? Well we use common sense and spend time to learn them through interacting with the community. Most contributors work to get ingrained into a community through helping with small PRs, helping with bugs (responding, labeling, fixing, etc), being present on IRC/matrix. By integrating they can learn the rules/processes and then work on moving and shaking on the larger pieces. @jvoisin is a shining example of this. You want to see someone who went to nobody to earning the trust to get added as a core member? jvoisin did this through integrating with the community and as a result is now the number 2 github contributor (Side note don't construe volume here as a direct measure of value - many other core folks contribute in ways that are arguably even more valuable).

What you've done so far is the opposite. Dump a bunch of large PRs, and not taken the time to get ingrained or learn the rules. Furthermore you've ran around and caused a bunch of drama by pushing buttons through large opinionated comments. My tone here suggests that these things are negative. But they don't have to be. I'm okay with pushing boundaries through large PRs and opinionated comments, but these things have to be balanced by equal efforts to the smaller more mundane aspects of getting ingrained into a community. How exactly you do so is up to you.

randomnicode commented:
> I understand the implications. The work is the sunk cost here. Just because you stacked those PRs on top of one another will not bias on if/how/when the base PR(s) get merged. I'm not trying to be a jerk, but rather this is a result of this community being very iterative on doing work. Whether that's good/bad/not inline with other FOSS communities can be debated separately as its irrelevant.

I actually don't know what you're talking about. I really don't know what your software background is, but collaborative and iterative work aren't mutually exclusive. In fact, most of these tools you use (like git) are designed for exactly addressing and solving those issues. Sunk cost is a ridiculous term being applied here when _you_ are _creating_ extra cost/work here that isn't sunk. In fact the entire argument here is why it needs to be sunk. This isn't about larger ideals of FOSS, and I'm not even sure why that's coming up. This is about the sheer common sense. Many different communities converge on similar solutions because they make sense, and they become best practices. What you're quoting to me just sounds an obstinate I-want-to-merge-my-PR-first-no-matter-what-effects-it-causes out of hubris or whatever unknown reason.

I really take it you haven't worked with a lot of large projects, because most of the stuff I've mentioned in the past few weeks would not be news to most software engineers. They'd be considered standard practices.

> You're assuming here that your PRs are going to get merged at all. In addition if your PRs are nicely contained it should be straightforward to get them updated with either a rebase or merge.

Okay, what? What do "nicely contained PRs" have to do with merges and rebases? You do understand that stacked PRs are building on top of each other, right? If I had made additional stacked PRs to split issues even more granularly by using whatever definition of "nicely contained" you have, all that would've accomplished is that I would've needed to merge or rebase even MORE PRs. That's how stacked PRs work. Please read a git primer. That will additionally get you out of the habit of _rebasing_ your public PRs that were used for "collaboration" instead of _merging_ master on top of it. I don't know about you, but how exactly _do_ you collaborate across PRs? Force everyone else to rebase all the time?

> I've updated this PR because your PR contains changes that I do not want to merge (either in the same PR, or even at all). In addition you've given me the impression you're not going to squash your changes. Am I being a meanie for making people break up their PRs into bite size chunks? Probably. However it is done for some very good reasons. Mainly 1. Allows easily reverting. 2. Makes for readable git history. 3. Makes bisecting issues much easier.

Dude, seriously, read a git primer. You're not going to squash commits on stacked PRs, that's rewriting history. It'll cause merge conflicts on merge conflicts. You'll have to resolve conflicts from PR1 on PR2, from PR1 (again!) AND PR2 on PR3, and so on, it grows exponentially the more PRs are stacked. You seriously don't seem to understand the basics here. Squashing might work for singular finished branches for the sake of keeping git history clean, but otherwise squashing and collaborating in parallel are just shooting yourself in the foot. And doubly so when you're asking people to submit small PRs. How else are they going to present an entire branch's worth of idea other than stacking PRs? Meanie doesn't begin to describe it. That's just plain obstinacy and not utilizing the tools at your disposal for some unknown rationale. Git history "cleanliness" being extremely subjective aside (I actually prefer to not lose commits as atoms of granular work, and actually be able to, you know, search through a repo's history instead of someone's reflog), you can actually revert commits too, not just PRs. No, really, try it. `git revert`.  You probably don't revert commits too much because you squash them. You're right, squashing commits isn't going to happen with these PRs. You took that away as soon as you wanted PRs broken up.

> We are going to work incrementally. Maybe this is all we are going to work on for spring boot right now. That sounds crazy but you have to realize there is no product manager, no full-time dev team committed to seeing a large change through end to end. Maybe you or I just vanish tomorrow and to never finish and never help the end users with bugs that surface as a result of a half-done large spring boot migration. Therefore we work incrementally and cautiously.

Yeah, that does sound crazy! I'm not sure why you're asking for "bold" contributions then if all you're going to do is sit on their PRs/effort while it grows stale or go "yeah, maybe in a year we'll think about it"? Why should people be contributing to you if their effort isn't valued or considered?

And if someone vanishes and leaves behind a ton of issues, then revert. You're using git. How many times do I have to restate this? Or fix the issues. You are being gifted someone's time and effort. May not be perfect or bug free, but it's a start. You're not the only open source project around, everyone has similar issues. Collaborators come and go for all software projects, open source and organizations etc. This isn't an airsonic problem. They seem to deal with it okay.

Oh, and GET DEV BUILDS. I don't understand your phobia of intermediate builds DESPITE the plethora of issues it solves for you and DESPITE someone even having done most of the leg work and having submitted a PR for it to start the process. And your response? "we'll just let it sit idle for a bit".

> You mean cause more work for you? PRs that are not clean, self contained and squashed cause more work for everyone else.

Yes, actually, it does include me, particularly considering I'm the author of the PR here. What part of my PR isn't "clean"? Especially considering this PR looks more or less identical? And especially considering my repeated mentions how similar this PR was going to turn out to #1364 the more you worked on it and duplicated effort (which you did anyway, and it ended up similar just as I said).
And I really don't understand your point about squashing here. You're just plain wrong. Again, read about git please. You're losing history with a squash and actually making things difficult for everyone. You won't be able to find any intermediate commits anymore that you can branch off of or revert if you squash stuff. You're making collaboration harder. The only place those commits will exist is in the reflog. Seriously, friend, I'm not sure why I'm explaining git here.

> So look here. At the end of the day I like your intentions. 

I'm not sure why this is worth mentioning. As opposed to not liking my intentions? I'm really indifferent. I'm not looking to be in someone's good books or to have my intentions praised. This is simply irrelevant.

> So what do we do without a pair of rules? Well we use common sense and spend time to learn them through interacting with the community. Most contributors work to get ingrained into a community through helping with small PRs, helping with bugs (responding, labeling, fixing, etc), being present on IRC/matrix. By integrating they can learn the rules/processes and then work on moving and shaking on the larger pieces. @jvoisin is a shining example of this. You want to see someone who went to nobody to earning the trust to get added as a core member? jvoisin did this through integrating with the community and as a result is now the number 2 github contributor (Side note don't construe volume here as a direct measure of value - many other core folks contribute in ways that are arguably even more valuable).

Yeah, no thanks. I'm not interested in joining what sounds very much like a clique. I came here to contribute. I don't know why @jvoisin was name dropped here (hey, good for him you view him highly, and from my limited interaction with him I think he's fine too!), but it doesn't really matter. I'm not interested in following his path to airsonic superstardom, or really anyone's path. What do you think I'm here for? Airsonic Hall of Fame?

I'm not sure what "most" contributors should "start off" doing per your dictates, but sorry, I'm simply not interested in any of that. Or I might be at a later point, but that's really up to me. I want to and will work on things I'm interested in (and that's not particular to this project either). If it means bigger stuff, then it's bigger stuff. If it strikes my mood to do smaller stuff, then it's smaller stuff. If it means fixing bugs, then it's fixing bugs. I'm not really looking for people to tell me what to do unless I ask. I'm contributing on my terms, not yours. Other folks can certainly ask me for help, and I can try, but ingratiating myself to a social community as a prerequisite to getting my code merged isn't something I'm interested in. Contributors come from all places with all different intentions and different backgrounds and it's silly to discount their effort because they don't want to join IRC/matrix/some-social-group, or help with smaller PRs first. I really don't care about earning your trust or really anyone's, my code speaks for itself. Or getting added as any kind of core member. This isn't a power trip.

I'm looking to try and make things more efficient so things can move along much faster and others can enjoy things that I have a skillset for before I have to depart again. My code is given as a contribution as-is (unless I want to change it), if you don't want it, I can just fork and keep it for everyone else. Being number 1 or number 2 contributor on this repo isn't a target, aspiration nor a concern of mine.

> What you've done so far is the opposite. Dump a bunch of large PRs, and not taken the time to get ingrained or learn the rules. Furthermore you've ran around and caused a bunch of drama by pushing buttons through large opinionated comments. My tone here suggests that these things are negative. But they don't have to be. I'm okay with pushing boundaries through large PRs and opinionated comments, but these things have to be balanced by equal efforts to the smaller more mundane aspects of getting ingrained into a community. How exactly you do so is up to you.

Look, my "opinionated" comments come from having particular skillsets and aptitude, and significant experience. And an impression that I can help with certain best practices not being followed that might just make lives easier for everyone and expedite feature rollouts. It is meant as help. Plenty of stuff I've observed here strikes me as stuff that junior engineers or folks with no formal software background would do. I thought I could help.

I'm not here to wage war against everyone who doesn't use the tools properly. You don't want to listen to my words? That's fine, you can continue going down whatever way you were going. It's certainly not affecting my life for trying to help. I'm definitely not running around. And I'm also most certainly not looking for you to be "okay" with aspects of things I say or do. Your approval isn't something I'm looking for nor care about.

Like I said, I'm not looking for a social outlet. My anonymity as a contributor in contributing to this repo is important to me. Being ingrained into any community is a slow process that might come with time organically, but I also don't really care to kickstart by being on IRC or adhering to some unwritten norms. I really don't have the time nor the inclination to learn these processes or whatever you have in mind. Sounds more like a clique to me. I came here to contribute. My reasons for contributions are mine alone and I don't really ascribe to and therefore don't really adhere to any of your prerequisites as preconditions. Yeah, no I don't want to label. Or fix small bugs. Or respond to issues. Or maybe I do. That's really not up for you to dictate. I'm going to work on whatever I feel like working on. And if my work is conditional on accomplishing some mundane tasks as proving my worth first, then I'd rather just not. My self worth isn't affected by this at all.

> How exactly you do so is up to you.

Right, all you've seem to have really done is just alienated me as a contributor. I would have preferred to stay on this fork and not split the community further so ongoing progress can stay concentrated and consolidated on one repo, but it seems that's not really an option with this fork or even one of it's goals. So here is what I think I'll do:
- close my PRs here
- fork this repo as airsonic2 or airsonic-advanced or whatever strikes my fancy
- merge my stuff there so it's more current than this fork at least.

I'll be on my merry way and bid you good luck in your future endeavors. I'll post a link to the new forked repo with more recent and advanced functionality here once it's done.

randomnicode commented:
> Maybe this is all we are going to work on for spring boot right now.

@fxthomas Thanks for the efforts. It was fun collaborating, I hope we have an opportunity to do that again. Doesn't seem like Spring boot 2 is even in the cards for now, sorry for your wasted time. if you want a more updated version of a subsonic/airsonic fork with all these goodies merged, I'll post it to you once I have a proper repo set up
